File: rust/src/problems/s1394_find_lucky_integer_in_an_array.rs
struct Solution;

impl Solution {
    pub fn find_lucky(arr: Vec<i32>) -> i32 {
        let mut map = std::collections::HashMap::new();
        let mut lucky_num = -1;
        for num in arr {
            let count = map.entry(num).or_insert(0);
            *count += 1;
        }
        for (num, count) in map.into_iter() {
            if num == count {
                lucky_num = lucky_num.max(num);
            }
        }
        lucky_num
    }
}

struct Example {
    input: Vec<i32>,
    output: i32,
}

#[test]
pub fn test() {
    let examples = vec![
        Example {
            input: vec![2, 2, 3, 4],
            output: 2,
        },
        Example {
            input: vec![1, 2, 2, 3, 3, 3],
            output: 3,
        },
        Example {
            input: vec![2, 2, 2, 3, 3],
            output: -1,
        },
    ];
    for example in examples {
        assert_eq!(Solution::find_lucky(example.input), example.output);
    }
}

File: typescript/src/problems/s1394_find_lucky_integer_in_an_array.ts
/**
 * @link https://leetcode.com/problems/find-lucky-integer-in-an-array/
 * @param arr
 */
export function findLucky(arr: number[]): number {
  let lucky = -1;
  // let prefix = -1;
  let count = 1;
  arr.sort((a, b) => a - b);
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] === arr[i - 1]) {
      count += 1;
    } else {
      if (arr[i - 1] === count) {
        lucky = arr[i - 1];
      }
      count = 1;
    }
    if (i === arr.length -1 && count === arr[i]) {
      lucky = arr[i];
    }
  }
  // const map = new Map<number, number>();
  // for (const num of arr) {
  //   const count = (map.get(num) ?? 0) + 1;
  //   map.set(num, count);
  //   if (count === num && num > lucky) {
  //     prefix = lucky;
  //     lucky = num;
  //   } else if (count !== num && lucky === num) {
  //     lucky = prefix;
  //   }
  // }
  // for (const num of arr) {
  //   if (map.get(num) === num) {
  //     lucky = Math.max(lucky, num);
  //   }
  // }
  return lucky;
}
